### PR TITLE
[PasswordHasher] UserPasswordHasher only calls getSalt when method exists

### LINKED ISSUE #41753: [Security] UserPasswordHasherInterface requires getSalt(), not on PasswordAuthenticatedUserInterface
**Symfony version(s) affected**: 5.3.2 (probably since 5.3.0)

**Description**
When using the `UserPasswordHasherInterface` with a User implementing the `PasswordAuthenticatedUserInterface`, but not the `UserInterface` you might run into the following error:

```
Attempted to call an undefined method named "getSalt" of class "<your user class>"
```

**How to reproduce**
1. Create a user implementing the `PasswordAuthenticatedUserInterface`:
    ```php
    namespace App\Entity;

    use Symfony\Component\Security\Core\User\PasswordAuthenticatedUserInterface;

    class User implements PasswordAuthenticatedUserInterface
    {
        private string $email;
        private string $password;

        public function __construct(
            private string $email,
            private string $password = '',
        ) {}

        public function getEmail(): string
        {
            return $this->email;
        }

        public function getPassword(): ?string
        {
            return $this->password;
        }

        public function updatePassword(string $hashedPassword): void
        {
            $this->password = $hashedPassword;
        }
    }
    ```

2. Create a new user and call the password hasher:
    ```php
    namespace App\Security;
    
    use App\Entity\User;
    use Doctrine\ORM\EntityManagerInterface;
    use Symfony\Component\PasswordHasher\Hasher\UserPasswordHasherInterface;
    
    final class UserCreator
    {
        public function __construct(
            private EntityManagerInterface $entityManager,
            private UserPasswordHasherInterface $passwordHasher,
        ) {}
    
        public function create(string $name, string $password): void
        {
            $user = new User($name);
            $user->updatePassword($this->passwordHasher->hashPassword($user, $password));
    
            $this->entityManager->persist($user);
            $this->entityManager->flush();
            $this->entityManager->clear();
        }
    }
    ```

**Actual Behavior**
Fails at [UserPasswordHasher#L46](https://github.com/symfony/symfony/blob/v5.3.2/src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php#L46), because `getSalt()` does not exist

**Expected Behavior**
`getSalt()` should not be called, when the method is not provided.

**Possible Solution**
Use `method_exists($user, 'getSalt')` before calling the method.

### PR DESCRIPTION
**Pull request by dbrumann**
| Q             | A
| ------------- | ---
| Branch?       | 5.3
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #41753
| License       | MIT

### CODE REVIEW

File: src/Symfony/Component/PasswordHasher/Tests/Hasher/UserPasswordHasherTest.php [hunk 1]
```diff
@@ -72,25 +71,43 @@ public function testHash()
...
 
         $passwordHasher = new UserPasswordHasher($mockPasswordHasherFactory);
 
-        $encoded = $passwordHasher->hashPassword($userMock, 'plainPassword');
+        $encoded = $passwordHasher->hashPassword($user, 'plainPassword');
         $this->assertEquals('hash', $encoded);
     }
 
-    public function testVerify()
+    public function testHashWithPasswordAuthenticatedUser()
```
Comment (dbrumann) [inline src/Symfony/Component/PasswordHasher/Tests/Hasher/UserPasswordHasherTest.php:81]: I will remove `expectError()` with the fix. This is just to show the bug and that replacing other user mocks with concrete instance will not affect their behavior to verify the fix will not affect them.

**Comment by chalasr** [comment]
Thanks for the perfect report and the patch, Denis.
Could you please check other places where the very same deprecation notice is triggered?
I suspect that at least `CheckCredentialsListener` and `PasswordMigratingListener` have the same issue (might be a few more)

**Comment by derrabus** [comment]
If a user class already implements `PasswordAuthenticatedUserInterface`, but not `LegacyPasswordAuthenticatedUserInterface`: do we still want to check if `getSalt()` exists?

**Comment by chalasr** [comment]
Very good question. 
When introducing the PasswordHasher contracts, I decided to not bind them to `UserInterface` to allow for greater flexibility for both developers and us (maintainers, for potential future evolutions).

But, the current code (before this patch) assumes that if your user class is not implementing `LegacyPasswordHasherInterface`, it still inevitably has the `getSalt()` method anyway because it is probably implementing `UserInterface` which enforces having this method until 6.0.

This patch basically removes that assumption. And actually, because `UserPasswordHasher` is located in password-hasher and not in security-core, it makes sense to me not to tie it to `UserInterface`. 
So this `method_exists()` check is just a way to not check for `$user instanceof UserInterface`. 
Yet it won't be needed at all in 6.0+, but having it now allows for non-`UserInterface` implementations to be compatible with this service, while they get a meaningful notice telling them to implement the new interface. 

We could consider this a too edge case for supporting it. 
Personally, I think it does not harm so I'm in favor of doing this change.

**Comment by dbrumann** [comment]
> Thanks for the perfect report and the patch, Denis.
> Could you please check other places where the very same deprecation notice is triggered?
> I suspect that at least `CheckCredentialsListener` and `PasswordMigratingListener` have the same issue (might be a few more)

Sure thing, but probably not today.

**Comment by dbrumann** [comment]
@chalasr as far as I can tell, the other places do not (yet) need this change, because all occurrences I could find where getting the user via `UserPassportInterface` which will always return a `UserInterface` which still has a `getSalt()` method.

**Comment by chalasr** [comment]
Thanks for confirming.

**Review by chalasr — APPROVED** [review]

File: src/Symfony/Component/PasswordHasher/Tests/Fixtures/TestLegacyPasswordAuthenticatedUser.php [hunk 2]
```diff
@@ -0,0 +1,53 @@
+<?php declare(strict_types=1);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/PasswordHasher/Tests/Fixtures/TestLegacyPasswordAuthenticatedUser.php]: strict types should be removed everywhere in the PR

File: src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php [hunk 3]
```diff
@@ -43,9 +43,12 @@ public function hashPassword($user, string $plainPassword): string
             trigger_deprecation('symfony/password-hasher', '5.3', 'The "%s()" method expects a "%s" instance as first argument. Not implementing it in class "%s" is deprecated.', __METHOD__, PasswordAuthenticatedUserInterface::class, get_debug_type($user));
         }
 
-        $salt = $user->getSalt();
-        if ($salt && !$user instanceof LegacyPasswordAuthenticatedUserInterface) {
-            trigger_deprecation('symfony/password-hasher', '5.3', 'Returning a string from "getSalt()" without implementing the "%s" interface is deprecated, the "%s" class should implement it.', LegacyPasswordAuthenticatedUserInterface::class, get_debug_type($user));
+        $salt = null;
+        if (method_exists($user, 'getSalt')) {
```
Comment (stof) [inline src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php]: If we keep support for that case (see my comment on the issue), I'd rather rewrite the code like that:

```php
$salt = null;

if ($user instanceof LegacyPasswordAuthenticatedUserInterface) {
    $salt = $user->getSalt();
} elseif ($user instanceof UserInterface) {
    $salt = $user->getSalt();

    if ($salt) {
        trigger_deprecation(...);
    }
}
```

this way, the code will support `getSalt` either for the new LegacyPasswordAuthenticatedUserInterface or when using UserInterface (i.e. the pre-5.3 API) and the code will be more similar to the 6.0 code (which has this `if` check but not the `elseif`) which simplifies merging branches

Comment (chalasr) [reply src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php]: Works for me.

Comment (dbrumann) [reply src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php]: done

**Review by derrabus — APPROVED** [review]

**Comment by derrabus** [comment]
Good catch, thanks Denis.

**Pull request merged**